[autoscaler] Flag flip for resource_demand_scheduler should take into account queue

AmeerHajAli opened the pull request:
I set the max back log size from 1 to 1M and it did not seem to affect much the performance, I set it now to 10k, we can change that in the future if necessary.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

wuisawesome commented:
Would you mind adding some sort of description about what autoscaling behavior now looks like? 

For example, how many nodes start up (and when) to execute 

```
f = ray.remote(sleep)
[f.remote(0.1) for _ in range(1000)]
```

and for 
```
f = ray.remote(sleep)
[f.remote(10) for _ in range(1000)]
```

ericl commented:
> I set the max back log size from 1 to 1M and it did not seem to affect much the performance, I set it now to 10k, we can change that in the future if necessary.

Can you also clarify how you measured "performance" here? It is not specified in the description.

AmeerHajAli commented:
Great questions.
I tested some examples like the ones @wuisawesome posted and I measured the time the remote calls finish and the time the get calls finish and they seemed to differ very slightly. I will report the timings Alex asked.

ericl commented:
One metric (unit test) I am interested in is say you have 100 nodes, and 10k tasks, how long does it take to do bin packing.

ericl commented:
One metric (unit test) I am interested in is say you have 100 nodes, and 10k tasks, how long does it take to do bin packing.

Intuitively bin packing is O(m*n), so 100 * 10000 might be quite slow.

AmeerHajAli commented:
The tests I had were for nano benchmarks that took a few minutes to run. The overhead @ericl mentions (is within seconds < update_interval_s) but is negligible compared to the benchmark. Perhaps it is best here to put a table (logarithmic increments) with different queue lengths (1-1M), number of nodes (1-10), tasks (1k-1M) and measure the total time spent on bin packing, remote calls and get calls. I set the number of max nodes to 10 because I am actually running this on AWS and don't want it to cost us a fortune. But 100k tasks with 10 nodes should be somewhat similar to 10 nodes and 100k tasks (with queue of 100k).

wuisawesome commented:
@AmeerHajAli I'm not sure how your micro benchmarks measure the bin packing time. The autoscaler isn't on task execution's critical path.  

An equivalent question here is: how high does the max size need to be before the binpacking takes 5s?

ericl commented:
@AmeerHajAli the macro scale testing on real clusters is good validation, but what I'm talking about is a unit test where we can easily scale up to hundreds of nodes.

What I care about most, is validating that this change doesn't fundamentally introduce too much overhead at the component level due to O(n*m) algorithm scaling. If we know this is fine, it is highly likely it will work in "real life".

AmeerHajAli commented:
I see. Will you guys be satisfied with a test that makes sure that the resource demand scheduler overhead is within a margin less than update_interval_s for say 1000 nodes.

AmeerHajAli commented:
Here are two data points:
```
backlog_size = 10k
number of nodes = 2k (1k cpu nodes and 1k gpu nodes)
demands = [{"GPU": 1}, {"CPU":1}] * (backlog_size=10k)
```
- The time `get_nodes_to_launch` takes is 4.6 seconds.
- Reducing `backlog_size` to 1k makes it take 0.32 seconds.
- Reducing the number of nodes to 200 (100 cpu nodes and 100 gpu nodes) makes it take 1.02 second.
- Reducing  `backlog_size` to 1k  and the number of nodes to 200 (100 cpu nodes and 100 gpu nodes) make it take 0.06 seconds.

Bottom line, If we want to be on the safe side (always less than 1 second) with 100-1000 nodes, we can set the backlog size to 1k. I also prefer this option since it is also less aggressive.

ericl commented on python/ray/tests/test_resource_demand_scheduler.py:
```diff
@@ -365,6 +366,48 @@ def test_calculate_node_resources():
+            utilizations[gpu_ips[i]] = {"GPU": 8, "CPU": 32}
+        assert len(utilizations) == 2 * num_nodes
+        t1 = time.time()
+        to_launch = scheduler.get_nodes_to_launch(all_nodes, {}, demands,
```
Mind asserting the value of to_launch here? We should ensure a big number here, to ensure we cover code to launch nodes as well.

wuisawesome commented on python/ray/ray_constants.py:
```diff
@@ -73,6 +73,10 @@ def env_bool(key, default):
 
+# The backlog queue size upper threshold to report for the demand based
+# autoscaling.
+MAX_BACKLOG_SIZE = 1000
```
Can we move this to `autoscaler/_private/constants.py`?

wuisawesome commented on python/ray/monitor.py:
```diff
@@ -49,7 +49,11 @@ def parse_resource_demands(resource_load_by_shape):
+                    # Limit the backlog queue size to reduce the overhead of
+                    # the resource demand scheduler.
+                    min(ray_constants.MAX_BACKLOG_SIZE,
+                        resource_demand_pb.backlog_size)):
```
Can we do `MAX_BACKLOG_SIZE - len(backlog_queue)` to bound the entire vector rather than the just the portion from the backlog?

AmeerHajAli replied:
Can we actually not? Since it is reduced to 1k, it should handle up to say 20 different types. I am worried this might fill the demand with the first type (say just CPUs) and sometimes it is better to allow a mix of CPUS and GPUS.

AmeerHajAli replied:
do you mean `MAX_BACKLOG_SIZE - len(waiting_bundles + infeasible_bundles)`?

AmeerHajAli commented:
@ericl @wuisawesome , I added a couple of more tests to see different resource demand vector sizes and different available nodes. Note that I changed the monitor function not to just trim the backlog bundles but also the `waiting_bundles + infeasible_bundles` So that we really bound the execution time of `get_nodes_to_launch`.

wuisawesome reviewed (changes requested):
Hey we need to block on merging this until we make sure we aren't scaling to aggressively. I will describe this in more detail in my simulator PR.

ericl commented:
> Hey we need to block on merging this until we make sure we aren't scaling to aggressively. I will describe this in more detail in my simulator PR.

Do you mean beyond the upscaling throttling? (e.g., launch at most 1 instance starting off, then at most 20% pending launches)

wuisawesome commented:
This may actually need some investigating (there's a chance this is a bug in the simulator still). I will get back to y'all on this by EOD.

AmeerHajAli commented:
Isn't that handled in the max launch concurrency?

ericl commented on python/ray/tests/test_resource_demand_scheduler.py:
```diff
@@ -365,6 +367,108 @@ def test_calculate_node_resources():
+    # is requires adding another ~100 nodes.
+    to_launch = test_backlog_queue_impact_on_binpacking_time_aux(
+        num_available_nodes=0,
+        time_to_assert=0.4,
```
Small timeouts are likely to cause flaky tests if the machine happens to be busy--- how about we raise to 10s as a sanity check, and just document how long it should take?

ericl commented:
Looks good, but let's raise the timeouts to prevent possible test flakiness.
